### ISSUE TITLE
Why using Jacobi decoding? What are the advantages besides the fact that Jacobi decoding can reduce some of the decode steps？

### ISSUE THREAD

**Issue by ZipECHO** [hao-ai-lab/LookaheadDecoding#33]
What are the advantages besides the fact that Jacobi decoding can reduce some of the decode steps？Can it be replaced by normal greedy decoding? How much negative impact will this have on the efficiency of decoding?

**Comment by shermansiu** [comment]
I'm not one of the authors, but I'll answer.

The reason we would apply Lookahead Decoding in the first place is that it can reduce the number of decoding steps, at the cost of extra FLOPS. Faster decoding is always nice.

Yes, of course. But it might be slower than LADE with the optimal settings for your GPU (assuming you have enough spare FLOPS to get a speedup out of LADE).

As for the negative impact, it depends on your GPU, the LADE settings, and the number of FLOPS needed to run your LLM. In general, you'll get more of a benefit from using LADE if you have a powerful GPU (e.g. a A100) as opposed to a 3090 or a V100.

**Comment by shermansiu** [comment]
The way to do this would be to find the optimal parameters experimentally for multiple models sizes for several GPUs and formulate a statistical scaling law for the optimal lookahead decoding parameters at each level.

I don't have the $$$ needed for this, unfortunately.

**Comment by jivanph** [comment]
I meant something more simple, like computing Flops without and with LookAhead for some fixed parameters. Doesn't have to be the optimal parameters but any parameters that show a faster decoding.

**Comment by shermansiu** [comment]
Without lookahead, would something like this work? https://blog.eleuther.ai/transformer-math

**Comment by shermansiu** [comment]
I think getting the estimated speedup given the model size, batch size, GPU VRAM, and GPU TFLOPS, etc. would be more useful though, as it's hard to have an intuitive understanding of the impact of FLOPS.

**Comment by jivanph** [comment]
What you mention surely would be more useful but also more demanding and even beyond the scope of a comparison for a fixed setup and fixed parameters.

**Comment by shermansiu** [comment]
Using the above transformer FLOPs equation, the total number of floating point operations would be roughly $2PD$ for a forward pass, where $P$ is the total number of parameters and $D$ is the (inference) dataset size.

**Comment by shermansiu** [comment]
For lookahead decoding, it seems that the lookahead branch uses $1+(N-1)W$ tokens for the lookahead branch and $(N-1)G$ tokens for verification, where

- $N$ is the n-gram size
- $W$ is the window size
- $G$ is the guess set size/number of candidates to consider

So, the total number of FLOPs would be about
$2PD(1+(N-1)(W+G))$.

**Comment by shermansiu** [comment]
Note that the above equations assume that inference is done at 16 bytes of precision.

**Comment by jivanph** [comment]
One of the things that caught my eye when I first looked at LookAhead is that "[LookAhead decoding] Linearly decreases #decoding steps relative to log(FLOPs) used per decoding step".

Did you guys @Viol2000 run experiments where you counted FLOPs? I ask because I was wondering how the log comment came to be.

**Comment by Viol2000** [comment]
We do not count FLOPs carefully; we just use #input tokens to approximate the per-step FLOPs. See this figure to find the relation between the log(FLOPs) and the compression ratio: https://lmsys.org/images/blog/laattention/match-scaling.png. I have a formulation in the paper to explain where the log operator comes from. Stay tuned!

**Comment by Viol2000** [comment]
Hi @shermansiu , thanks for your careful counting of the per-step FLOPs. Transformers' FLOPs mainly come from the attention and the mlp layer. mlp layers' FLOPs are proportional to the # input tokens. And attention layers' FLOPs are also proportional to the # input tokens to some extent. So I just use # input tokens to approximate the per-step FLOPs (i.e., if you input 241 tokens, you roughly have 241x larger FLOPs.)

**Comment by shermansiu** [comment]
Actually, with a KV-cache, increasing the number of input tokens alone by $\delta$ should not cause the number of FLOPs to increase by a factor of $\alpha$ times (multiplicatively). Normally, that would just cause the inference cost to increase by $2P\delta$ (plus the correction factor above), additively, according to the OpenAI scaling laws. i.e. New cost $\approx$ old cost + $2P\delta$.

The problem is that the OpenAI compute estimate discards the context-dependent terms, because "Since we primarily study models where $d_{model} \gg n_{ctx}/12$, we do not include context-dependent terms in our training compute estimate." Interestingly enough, they also do not include the cost for the self-attention computation in Table 1 of their scaling laws paper. (The different notation comes from the OpenAI scaling laws paper).

The reason we *also* have a multiplicative increase (on top of the additive increase) is because the lookahead decoding part is re-computed each Jacobi decoding step and can't be KV-cached.

---

As an aside, I think the FLOPs attention mask term in the OpenAI scaling laws paper is wrongly attributed... The $2n_{layer}n_{ctx}n_{attn}$ cost is not the cost of masking, but of multiplying $\mathrm{softmax}(\mathrm{mask}(qK^T))$, a $1\times n_{ctx}$ vector, by the $n_{ctx}\times d_{attn}$ $V$ vectors across all layers. Masking takes $n_{layer}n_{ctx}$ floating point instructions (not operations, which are more costly).

**Comment by shermansiu** [comment]
Here's an oversimplified, but more intuitive way to think about it, using the dimensions of the $qK^T$/$QK^T$. We are looking at inference per-token.

|  Situation  | Shape |
| ------------- | ------------- |
| No KV-cache  | $QK^T$: $(\ell+1)\times (\ell+1)$ |
| KV-cache | $qK^T$: $1\times (\ell+1)$  |
| Prompt longer by $\delta$, KV-cache | $qK^T$: $1\times (\ell + \delta+1)$  |
| Speculative decoding, KV-cache | $qK^T$: $(\alpha+1)\times (\ell+1 + \alpha)$  |

For the full generation, the +1 terms in the width (i.e. the second dimension) get added up to $L-\Lambda$.

**Comment by jivanph** [comment]
Thank you so much @shermansiu This discussion enlightened me. I'm still trying to understand the trade-off, and from the FLOPs sides I have a clearer picture now. Do you know from the code if there is an easy way to extract the total steps taken during LookAhead runs? I see it expressed in the summary but I don't know if there's a way to store this statistic as an output as well.

**Comment by shermansiu** [comment]
You mean something like saving it to a file or used elsewhere in your code?

The relevant line for the variable `steps` is here: https://github.com/hao-ai-lab/LookaheadDecoding/blob/b756db313419298d292a927c6dda950020ec1073/lade/decoding.py#L495

Use it however you wish.

**Comment by jivanph** [comment]
Thank you, I'm trying to store the steps variable and not just print it.

**Comment by shermansiu** [comment]
You'd need to modify the code for that.

The simplest (but hacky) way to obtain the value without having to modify a bunch of return calls is to have it call one of your own functions, say `append_steps` in line 496 of the aforementioned file.
